### ISSUE TITLE
Problem rsyncing with the latest release candidate

### ISSUE THREAD

**Issue by djmittens** [tknerr/bills-kitchen#91]
Hi, could you tell me what im doing wrong here ?

```
==> default:   - Exclude: [".vagrant/", ".git/", "node_modules", "bower_components"]
      1 [main] rsync 8052 find_fast_cwd: WARNING: Couldn't compute FAST_CWD pointer.  Please report this problem to
the public mailing list cygwin@cygwin.com
There was an error when attempting to rsync a synced folder.
Please inspect the error message below for more info.

Host path: /w/repo/erewhon/
Guest path: /src
Command: rsync --verbose --archive --delete -z --copy-links --chmod=ugo=rwX --no-perms --no-owner --no-group --rsync-path sudo rsync -e ssh -p 2222 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -i 'W:/repo/erewhon/cookbook/.vagrant/machines/default/virtualbox/private_key' --exclude .vagrant/ --exclude .git/ --exclude node_modules --exclude bower_components /w/repo/erewhon/ vagrant@127.0.0.1:/src
Error:       1 [main] rsync 8052 find_fast_cwd: WARNING: Couldn't compute FAST_CWD pointer.  Please report this problem to
the public mailing list cygwin@cygwin.com
Warning: Permanently added '[127.0.0.1]:2222' (ECDSA) to the list of known hosts.
rsync: change_dir "/w/repo/erewhon" failed: No such file or directory (2)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at /home/lapo/package/rsync-3.0.9-1/src/rsync-3.0.9/main.c(1052) [sender=3.0.9]
```

this is the line in my vagrant file responsible for rsync

```
  config.vm.synced_folder "..", "/src/", type: :rsync, rsync__exclude: [".git/", "node_modules", "bower_components"],
                          rsync__auto: true
```

**Comment by djmittens** [comment]
After doing a bit of googling it may be related to me running windows 8.1

http://stackoverflow.com/questions/22821467/cygwin-win-8-1-error-find-fast-cwd-warning-couldnt-compute-fast-cwd-pointer

Has anyone seen these / know how to fix em?

**Comment by tknerr** [comment]
@djmittens can you share your Vagrantfile? I have a Win7 machine here and could give it a try

**Comment by djmittens** [comment]
I think it will work with windows 7 no problem, i think 8.1 windows updated their libraries, and cygwin has a well, i think updating the binaries for cygwin-ssh and cygwin-rsync will solve this problem, but it probably works no problem as is on windows 7.

**Comment by tknerr** [comment]
Yep, the included rsync and ssh binaries are quite a bit aged...

**Comment by djmittens** [comment]
Unfortunately i dont know enough to update, however i can confirm that using rwRsync instead while doesnt work doesnt exhibit this problem,

```
There was an error when attempting to rsync a synced folder.                                                                                              
Please inspect the error message below for more info.                                                                                                     
                                                                                                                                                          
Host path: /w/repo/erewhon/cookbook/                                                                                                                      
Guest path: /src                                                                                                                                          
Command: rsync --verbose --archive --delete -z --copy-links --chmod=ugo=rwX --no-perms --no-owner --no-group --rsync-path sudo rsync -e ssh -p 2222 -o Str
ictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -i 'W:/repo/erewhon/cookbook/.vagrant/machines/default/virtualbox/private_key' --exclude .vagrant/ -
-exclude .git/ --exclude node_modules --exclude bower_components /w/repo/erewhon/cookbook/ vagrant@127.0.0.1:/src                                         
Error: Warning: Permanently added '[127.0.0.1]:2222' (ECDSA) to the list of known hosts.                                                                  
rsync: change_dir "/w/repo/erewhon/cookbook" failed: No such file or directory (2)                                                                        
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1165) [sender=3.1.1]                                         
```

**Comment by tknerr** [comment]
I'm preparing something for you to test, just checking whether it works on my win7...

**Comment by tknerr** [comment]
Hmmm... I get the same error with the current rsync / ssh version as you do with rwRsync:
```
There was an error when attempting to rsync a synced folder.
Please inspect the error message below for more info.

Host path: /c/Repos/_github/bills-kitchen/target/build/repo/vagrant-managed-servers/
Guest path: /src
Command: rsync --verbose --archive --delete -z --copy-links --chmod=ugo=rwX --no-perms --no-owner --no-group --rsync-path sudo rsync -e ssh -p 2222 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -i 'C:/Repos/_github/bills-kitchen/target/build/repo/vagrant-managed-servers/.vagrant/machines/fake_managed_server/virtualbox/private_key' --exclude .vagrant/ --exclude .git/ /c/Repos/_github/bills-kitchen/target/build/repo/vagrant-managed-servers/ vagrant@127.0.0.1:/src
Error: Warning: Permanently added '[127.0.0.1]:2222' (ECDSA) to the list of known hosts.
rsync: change_dir "/c/Repos/_github/bills-kitchen/target/build/repo/vagrant-managed-servers" failed: No such file or directory (2)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at /home/lapo/package/rsync-3.0.9-1/src/rsync-3.0.9/main.c(1052) [sender=3.0.9]
```

**Comment by djmittens** [comment]
thats a different problem it has to do with vagrant not running in bash, and when that happens it doesnt prepend with /cygdrive/, but also rsync needs the /cygdrive/ even when not in bash.

https://github.com/mitchellh/vagrant/issues/4073

**Comment by tknerr** [comment]
@djmittens I actually never tried using `config.vm.synced_folder ".", "/src", type: :rsync`

But I have this plugin working with rsync: 
https://github.com/tknerr/vagrant-managed-servers/blob/master/lib/vagrant-managed-servers/action/sync_folders.rb

Actually I did fix this too:
https://github.com/tknerr/vagrant-managed-servers/blob/master/lib/vagrant-managed-servers/action/sync_folders.rb#L58

Will use this plugin instead of `config.vm.synced_folder` for testing..

**Comment by djmittens** [comment]
is there any way we could just get the synced_folder working? i would rather not add more plugins to my setup since this cookbook has to work for mac windows and linux.

**Comment by tknerr** [comment]
@djmittens didn't mean to suggest to you to use a plugin for that ;-)

Would love to see synced_folder working, but have not digged through https://github.com/mitchellh/vagrant/issues/4073 yet. 

Can you try the files from https://github.com/tknerr/bills-kitchen/tree/update-ssh-rsync/files/tools and see if these fix the "Couldn't compute FAST_CWD pointer" warning for you?

**Comment by tknerr** [comment]
But I think we would still need to find a workaround for mitchellh/vagrant#4073. Too tired now, maybe you can give it a try?

**Comment by tknerr** [comment]
In a quick test the `ENV["VAGRANT_DETECTED_OS"] = ENV["VAGRANT_DETECTED_OS"].to_s + " cygwin"` workaround did not work for me

**Comment by djmittens** [comment]
thats what i opened the other ticket for

and closed it lol, it makes things worse.

Unless you package all of cygwin that wouldnt do you any good.

**Comment by tknerr** [comment]
Still not an ideal solution imho, as it requires to put this into every Vagrantfile:
`ENV["VAGRANT_DETECTED_OS"] = ENV["VAGRANT_DETECTED_OS"].to_s`

Alternatives:

 * put it in the global W:\home\.vagrant.d\Vagrantfile
 * patch vagrant after installation

**Comment by tknerr** [comment]
@djmittens added the `cygpath.exe` here:
https://github.com/tknerr/bills-kitchen/blob/update-ssh-rsync/files/tools/cygwin-rsync/cygpath.exe

For my primitive testing this was all that is needed in combination with the VAGRANT_DETECTED_OS workaround (added it to the global Vagrantfile):
https://github.com/tknerr/bills-kitchen/commit/328ba0ae298047e52f46f0f98732a1180b397073

**Comment by tknerr** [comment]
Maybe you can do some more testing if that works for all your cases. I will also need to do some basic testing before merging PR #92

**Comment by djmittens** [comment]
after downloading all of the files i get this result which is not what i would have wanted lol
```
There was an error when attempting to rsync a synced folder.
Please inspect the error message below for more info.

Host path: /w/repo/erewhon/cookbook/
Guest path: /src
Command: rsync --verbose --archive --delete -z --copy-links --chmod=ugo=rwX --no-perms --no-owner --no-group --rsync-path sudo rsync -e ssh -p 2222 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -i 'W:/repo/erewhon/cookbook/.vagrant/machines/default/virtualbox/private_key' --exclude .vagrant/ --exclude .git/ --exclude node_modules --exclude bower_components /w/repo/erewhon/cookbook/ vagrant@127.0.0.1:/src
Error: Could not create directory '/home/Nick/.ssh'.
Warning: Permanently added '[127.0.0.1]:2222' (ECDSA) to the list of known hosts.
rsync: change_dir "/w/repo/erewhon/cookbook" failed: No such file or directory (2)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1165) [sender=3.1.1]
```

Clearly that is not the directory it should be using

**Comment by djmittens** [comment]
After using cwRsync's ssh i was able to finally get the whole thing to work
```
==> default: Mounting shared folders...
    default: /tmp/vagrant-cache => W:/home/.vagrant.d/cache/chef/ubuntu-14.04
    default: /tmp/vagrant-chef/652ab00794841bbeb60a040f27628239/cookbooks => W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150409-5528-1v4w2sn-default
```

So now as soon as we fix the ssh thing i think its good to go.

**Comment by tknerr** [comment]
@djmittens just to be sure: did you replace the whole `cywin-ssh` and `cygwin-rsync` directories? or just the executables in it?

I got the home directory error too, but with the older version of the ssh and rsync exe, and this was without the `VAGRANT_DETECTED_OS` to be set.

Minimal Vagrantfile that works for me:
```
Vagrant.configure(2) do |config|
  config.vm.box = "chef/ubuntu-12.04-i386"
  config.vm.synced_folder "..", "/src/", type: :rsync, rsync__exclude: [".git/", "node_modules", "bower_components"], rsync__auto: true
end
```

Log:
```
C:\Repos\_github\bills-kitchen\target\build\repo\tmp>vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Box 'chef/ubuntu-12.04-i386' could not be found. Attempting to find and install...
    default: Box Provider: virtualbox
    default: Box Version: >= 0
==> default: Loading metadata for box 'chef/ubuntu-12.04-i386'
    default: URL: https://atlas.hashicorp.com/chef/ubuntu-12.04-i386
==> default: Adding box 'chef/ubuntu-12.04-i386' (v1.0.0) for provider: virtualbox
    default: Downloading: https://atlas.hashicorp.com/chef/boxes/ubuntu-12.04-i386/versions/1.0.0/providers/virtualbox.box
    default: Progress: 100% (Rate: 1986k/s, Estimated time remaining: --:--:--)
==> default: Successfully added box 'chef/ubuntu-12.04-i386' (v1.0.0) for 'virtualbox'!
==> default: Importing base box 'chef/ubuntu-12.04-i386'...
==> default: Matching MAC address for NAT networking...
==> default: Checking if box 'chef/ubuntu-12.04-i386' is up to date...
==> default: Setting the name of the VM: tmp_default_1428659712812_34566
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 => 2222 (adapter 1)
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: Warning: Connection timeout. Retrying...
    default:
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.
    default:
    default: Inserting generated public key within guest...
    default: Removing insecure key from the guest if its present...
    default: Key inserted! Disconnecting and reconnecting using new SSH key...
==> default: Machine booted and ready!
==> default: Checking for guest additions in VM...
==> default: Rsyncing folder: /cygdrive/c/Repos/_github/bills-kitchen/target/build/repo/ => /src
==> default:   - Exclude: [".vagrant/", ".git/", "node_modules", "bower_components"]
==> default: Mounting shared folders...
    default: /vagrant => C:/Repos/_github/bills-kitchen/target/build/repo/tmp
    default: /tmp/vagrant-cache => C:/Repos/_github/bills-kitchen/target/build/home/.vagrant.d/cache/chef/ubuntu-12.04-i386
==> default: Configuring cache buckets...
```

**Comment by djmittens** [comment]
It works, without the ssh, the way i copied those in is by checking out your pull request in a separate directory, then i completely emptied out cygwin-ssh folder, and i pasted in all of the files from your pull request, and it gave me that folder error.  When i renamed the folder to get out of my path,  which used  cwRsync's ssh it worked flawlessly.  So i am honestly not sure whats up with this, maybe its another windows 8.1 issue? The same thing happened when i installed cygwin, then copied file by file all of the files out of it and into this directory as well, that didnt work either, or put the local cygwin installation to the front of the path same issue, it works inside of the cygwin shell but not inside of bills-kitchen.

**Comment by tknerr** [comment]
@djmittens FYI after some more experimentation I switched to cwRsync now. Not only does it work for your case too, it seems to be a bit faster too. Also, instead of modifying the `VAGRANT_DETECTED_OS` environment, I decided to patch vagrant, so we don't need cygdrive.exe and it seems also a bit faster.

See the updated pull request here:
https://github.com/tknerr/bills-kitchen/pull/92

**Comment by tknerr** [comment]
fixed via #92 - should work for you now once the next rc release is out. if not, please reopen.

**Comment by djmittens** [comment]
Would you happen to know when the next rc is going to be getting released? Since im not quite sure how to build it myself to test this out.

**Comment by tknerr** [comment]
@djmittens waiting for docker 1.6.0 and chefdk 0.5.0 final versions to come out for inclusion in bills-kitchen 3.0

I might build another RC earlier, let me see if I can make it this week...

**Comment by tknerr** [comment]
@djmittens FYI - this is released in 3.0-rc2 now: 
https://github.com/tknerr/bills-kitchen/releases/tag/3.0-rc2

**Comment by djmittens** [comment]
@tknerr LGTM, i tried it out today, and everything seems to work allright.
